WordPress/src/main/java/org/wordpress/android/ui/media/services/MediaUploadService.java

```diff
@@ -100,75 +105,100 @@ public int onStartCommand(Intent intent, int flags, int startId) {
-        // event.media, event.error
+        MediaModel media = getMediaFromQueueById(event.media.getId());
+        if (media != null) {
+            media.setUploadState(UploadState.FAILED.name());
+            mDispatcher.dispatch(MediaActionBuilder.newUpdateMediaAction(media));
+        }
+        completeUploadWithId(event.media.getId());
         uploadNextInQueue();
     }
 
-    private void uploadNextInQueue() {
-        // waiting for response to current upload request
-        if (mCurrentUpload != null) {
-            AppLog.i(AppLog.T.MEDIA, "Ignoring request to uploadNextInQueue, only one media item can be uploaded at a time.");
-            return;
+    private synchronized void updatePostWithMediaUrl(MediaModel media){
+        if (media != null) {
+            PostModel post = mPostStore.getPostByLocalPostId(media.getPostId());
+            if (post != null) {
+                // actually replace the media ID with the media uri
+                MediaUploadReadyListener processor = new MediaUploadReadyProcessor();
+                PostModel modifiedPost = processor.replaceMediaFileWithUrlInPost(post, String.valueOf(media.getId()), FluxCUtils.mediaFileFromMediaModel(media));
+                if (modifiedPost != null) {
+                    post = modifiedPost;
+                }
+
+                // we changed the post, so let’s mark this down
+                if (!post.isLocalDraft()) {
+                    post.setIsLocallyChanged(true);
+                }
+                post.setDateLocallyChanged(DateTimeUtils.iso8601FromTimestamp(System.currentTimeMillis() / 1000));
+
+                // finally save the post
+                savePostToDb(post);
+            }
         }
+    }
 
-        // somehow lost our reference to the site, complete this action
-        if (mSite == null) {
-            AppLog.i(AppLog.T.MEDIA, "Unexpected state, site is null. Skipping this request - MediaUploadService.");
+    private synchronized void savePostToDb(PostModel post) {
+        mDispatcher.dispatch(PostActionBuilder.newUpdatePostAction(post));
+    }
+
+    private void uploadNextInQueue() {
+
+        MediaModel next = getNextMediaToUpload();
+
+        if (next == null) {
+            AppLog.v(AppLog.T.MEDIA, "No more media items to upload. Skipping this request - MediaUploadService.");
             stopServiceIfUploadsComplete();
             return;
         }
 
-        mCurrentUpload = getNextMediaToUpload();
+        SiteModel site = mSiteStore.getSiteByLocalId(next.getLocalSiteId());
 
-        if (mCurrentUpload == null) {
-            AppLog.v(AppLog.T.MEDIA, "No more media items to upload. Skipping this request - MediaUploadService.");
+        // somehow lost our reference to the site, complete this action
+        if (site == null) {
+            AppLog.i(AppLog.T.MEDIA, "Unexpected state, site is null. Skipping this request - MediaUploadService.");
```

Comment:
We should `return` here like we do in the `next == null` check above - there's nothing for `dispatchUploadAction` to do (and it'll fix the `'site' might be null` warning we're getting).